### ISSUE TITLE
Ubuntu 23.04 produces GRUB "error 452: out of range pointer" when booting in BIOS mode

### ISSUE THREAD

**Issue by pbatard** [pbatard/rufus#2233]
Well, it had to happen, with GRUB 2 **NOT** releasing in a timely manner, [distro developers have taken upon themselves to cherry pick GRUB patches](https://launchpad.net/ubuntu/+source/grub2/+changelog) with the end result that some of these BREAK compatibility between their kernel and a _vanilla_ GRUB 2.06 bootloader.

Case in point, Ubuntu 23.04 now produces a random `error 452: out of range pointer` message when trying to boot using our _vanilla_ GRUB 2.06 `core.img`.

Well, I don't have the time to sieve through this crap, so we're going to have to upgrade our GRUB 2.06 `core.img` to latest (since we confirmed it fixed the issue) and use that as our bootloader. This is most likely going to break boot for other ISOs, but hey, at least it'll keep the Ubuntu and GRUB folks **COMPLETELY OBLIVIOUS** to the bullshit they are generating for others...

**Comment by pbatard** [comment]
Fedora 38 also seems to be affected by the same issue, so we'll need to fix the `core.img` from https://github.com/pbatard/rufus-web/tree/gh-pages/files/grub-2.06-nonstandard...

**Comment by pbatard** [comment]
Aaaaand, __as expected__, fixing GRUB for 23.04 [BREAKs boot for older versions, such as 22.04](https://old.reddit.com/r/Ubuntu/comments/131xwg8/402_out_of_range_pointer_error_in_ubuntu/).

**Comment by pourdelir** [comment]
I got the same error.
On the same system (sand pit server) Ubuntu 22.1 was working fine.

,

**Comment by pbatard** [comment]
> I got the same error.

With what ISO? Can you please provide the full name of the ISO you used?

**Comment by pourdelir** [comment]
> > I got the same error.
> 
> With what ISO? Can you please provide the full name of the ISO you used?

Using Rufus 4.0p (portable) has resolved the issue. 
Ubuntu 23.04

**Comment by frizib** [comment]
A similar problem.

Rufus + Ubuntu Server 22.04 (the latest from the official website)

**Comment by pbatard** [comment]
Yes, that's what I said earlier when I mentioned that _fixing GRUB for 23.04 [BREAKs boot for older versions, such as 22.04](https://old.reddit.com/r/Ubuntu/comments/131xwg8/402_out_of_range_pointer_error_in_ubuntu/)._

Right now I don't have a good solution apart from advising people installing older versions of Ubuntu to use Rufus 3.22.

The problem is that GRUB 2.06 with patches is now incompatible with GRUB 2.06 without patches, so using a version of GRUB that works for 23.04 breaks 22.04. And the other problem is that, whereas Rufus does have a mechanism in place to download special version of GRUB as needed, this only works if the distro maintainers _do the right thing™_ and change the GRUB version __AS THEY SHOULD__ when they are no longer using the vanilla source.

In other words, if Ubuntu had edited their GRUB version so that it says something like `2.06-ubuntu-23` this whole mess could have easily been avoided, but instead, even though they are most definitely __NOT__ using GRUB 2.06, they did not alter the version, which means that both Ubuntu 22.x and Ubuntu 23.04 report that they are using GRUB `2.06`.

I also looked into trying to detect the version from the presence/absence of specific GRUB modules, but, sadly, there are no differences between 22.x and 23.x there. And I want a mechanism that works for more than Ubuntu, as Fedora suffers from a similar issue (Fedora 38 also requires a "patched" GRUB 2.06 to work, while also reporting that it uses plain 2.06), so I can't do something that's specific to detecting Ubuntu.

Right now, the solution I am being cornered into, both because the GRUB project really screwed __everyone__ with their lack of timely releases and because distro maintainers don't do the basic job of appending a custom suffix to the GRUB version the minute they stop using the original unpatched source, is use the ISO label as a GRUB 2.06 version suffix to try to differentiate between GRUB versions that need the recent out of range patch and those who don't, and then use the existing Rufus mechanism of downloading remote GRUB/Syslinux bootloaders. But even that is tricky because labels can contain tons of stuff (e.g. arch, X environment, "flavour") that we have to remove before we can get to something that doesn't require creating tons of symbolic link for each distro release, so I'm going to have to add some kind of ISO label sanitizer before I can get to a half-assed solution (since it will still require me to create symbolic links every time we discover a new release that requires a patched 2.06).

And the worst thing of all is that both the Ubuntu maintainers and, especially, the GRUB maintainers, have no idea of the complete shit show they created by not following the __FRIGGING ESTABLISHED BEST PRACTICES THAT EXIST FOR A VERY GOOD REASON__ of:
- Altering the version you report the minute you stop using the vanilla source from upstream (Ubuntu)
- __RELEASING EARLY, RELEASING OFTEN__ (GRUB).

And, unfortunately as usual, it is the end-users that pay the price of developers thinking that they are above following best practices, or thinking that they somehow _"know better"_ and that they somehow shouldn't apply to their projects...

**Comment by pbatard** [comment]
Alternatively, we may have to rely on [osinfo-db](https://gitlab.com/libosinfo/osinfo-db/-/tree/main/data/os) to tie a label regex to some kind of identifier we can use for GRUB.

Unfortunately, while osinfo-db is officially a database, it's one that spread into tons of small files and it doesn't look like it's designed to perform operations like `volume-id` &rarr; `short-id`, so we'd have to add our own custom parsing to generate something we can use (and therefore need to maintain) and the DB also isn't going to help us with things like `super_grub2_disk` and whatnot...

**Comment by pbatard** [comment]
Just a few notes for those who might still be experiencing this issue:
1. Please make sure that you use Rufus 4.1 or later.
2. If you created a drive with Rufus in the past, try <kbd>Alt</kbd>-<kbd>D</kbd> to force Rufus to download the relevant GRUB bootloaders again.
3. If this doesn't work and you are still getting the `error 452: out of range pointer`, please open a new issue in this tracker __and make sure to indicate what ISO you are using and where you downloaded it__.

**Comment by Nextinor** [comment]
I were using rufus 4.0p and my computer was unable to boot up with error 452: out of range pointer
using 4.1p it worked with no problem...
thank you @pbatard

**Comment by GreasyMonkee** [comment]
> Please make sure that you use Rufus 4.1 or later.
> If you created a drive with Rufus in the past, try Alt-D to force Rufus to download the relevant GRUB bootloaders again.

Sorry for the dumb question - the "ALT-D" sequence, is this applicable for a USB device that has been formatted and a complete new download of RUFUS 4.1?
How can this be applied using Windows 10 download of RUFUS - to make sure I am forcing a new GRUB download.

Reason I ask is that I am still getting "452 out of range" errors for Fedora Rawhide (FC39), downloaded from https://dl.fedoraproject.org/pub/fedora/linux/development/rawhide/Server/x86_64/iso/

Hardware is HPE DL380p Gen8 server, using Legacy BIOS - tried two different USB sticks, Sandisk 256GB, formatted as Large-FAT32

**Comment by pbatard** [comment]
<kbd>Alt</kbd>-<kbd>D</kbd> remove all the files previously downloaded by Rufus on the __host__ computer. It doesn't do anything to the target drive.

You will see a line at the bottom of Rufus (status bar) that says `Deleting directory 'C:\Users\<your_user_name>\AppData\Local\Rufus'` and once you have seen that line, then you can be sure that Rufus will not reuse any existing GRUB files you might previously have downloaded.

Now, I have just tested `Fedora-Server-netinst-x86_64-Rawhide-20230704.n.0.iso` and I can replicate your issue so, _as usual_, it appears that Fedora have applied yet some additional changes to their GRUB bootloaders since Fedora 37 1.7 (which I confirmed again to work, just in case), which broke boot.

Oh and since the Fedora folks __stupidly__ decided that their RawHide ISO label should not have any hint of a version (the one from the problematic ISO is simply `Fedora-S-dvd-x86_64-rawh`), and, because GRUB has been stuck in release-limbo for the past 2 years, we very much rely on distro maintainers providing a version so that we have a fighting chance of downloading the relevant GRUB files, I'm not sure I'm going to be able to do anything about the complete clusterfuck that GRUB/Fedora have imparted on mere users like yourselves, who just want a media that boots on BIOS systems.

For now, and as Rufus explicitly advises you to do in a prominent prompt every time you create a Fedora media, you should try creating the media in DD mode, as you should find it should work around the issue.

**Comment by GreasyMonkee** [comment]
@pbatard, many thanks for your quick response.

Indeed using DD worked, and the installation process began correctly.

I am not sure if it is related to the GRUB mess that has been created, but when I went to re-install my server (due to a newly broken install - thanks to DNF 5 and PHP upgrades gone wrong, that is a long story) from a ReaR backup image, I found that the existing three backups all failed due to some weird issues as it was trying to run systemd-udevd.service.

Now I am having to manually re-build the server from scratch, at least I can go to Rawhide.....

**Comment by pbatard** [comment]
Oh man, thanks to the lack of timely GRUB releases, Fedora currently have [332 patches](https://src.fedoraproject.org/rpms/grub2/blob/rawhide/f/grub.patches) (!) that they apply on top of GRUB 2.06.

I think I'm gonna wait for the official 2.12 GRUB release, that __may__ eventually happen sometime this year (but please don't ask me what happened to the 2.07 to 2.11 GRUB releases), and for Rawhide to switch to 2.12, because there's no way I'm gonna sieve through 332 patches to figure out which ones are needed to fix Fedora's out of range pointer error on BIOS...
